aquadoggo/src/graphql/client/dynamic_query.rs

```diff
@@ -0,0 +1,556 @@
+        // be a single document query if the schema name itself starts with `all_`).
+        if field_name.starts_with("all_") {
+            if let Ok(schema_id) = field_name.split_at(4).1.parse::<SchemaId>() {
+                // Retrieve the schema to make sure that this is actually a schema and doesn't just
+                // look like one.
+                let schema_provider = ctx.data_unchecked::<SchemaProvider>();
+                if schema_provider.get(&schema_id).await.is_some() {
+                    return self.query_listing(&schema_id, ctx).await;
+                }
+            }
+        }
+
+        // We now know that this is not a listing query. Continue by treating it as a single
+        // document query. Return `Ok(None)` if that doesn't work to signal that other resolvers
+        // should be tried for this field.
+        match field_name.parse::<SchemaId>() {
+            Ok(schema) => self.query_single(&schema, ctx).await,
+            Err(_) => Ok(None),
+        }
+    }
+}
+
+impl DynamicQuery {
+    /// Returns a single document as a GraphQL value.
+    async fn query_single(
+        &self,
+        schema_id: &SchemaId,
+        ctx: &Context<'_>,
+    ) -> ServerResult<Option<Value>> {
+        info!("Handling single query for {}", schema_id);
+
+        let document_id_arg = ctx.param_value::<Option<DocumentIdScalar>>("id", None)?;
+        let view_id_arg = ctx.param_value::<Option<DocumentViewIdScalar>>("viewId", None)?;
+
+        if let Some(view_id_scalar) = view_id_arg.clone().1 {
+            let view_id = view_id_scalar.try_into().map_err(|err: Error| {
+                ServerError::new(
+                    format!("Could not parse `viewId` argument: {}", err.message),
+                    Some(view_id_arg.0),
+                )
+            })?;
+
+            // If the `viewId` argument is set we ignore the `id` argument because it doesn't
+            // provide any additional information that we don't get from the view id.
+            return Ok(Some(
+                self.get_by_document_view_id(
+                    view_id,
+                    ctx,
+                    ctx.field().selection_set().collect(),
+                    Some(schema_id),
+                )
+                .await?,
+            ));
+        }
+
+        match document_id_arg {
+            (pos, Some(document_id_scalar)) => Ok(Some(
+                self.get_by_document_id(
+                    document_id_scalar.try_into().map_err(|err: Error| {
+                        ServerError::new(
+                            format!("Could not parse `id` argument: {}", err.message),
+                            Some(pos),
+                        )
+                    })?,
+                    ctx,
+                    ctx.field().selection_set().collect(),
+                    Some(schema_id),
+                )
+                .await?,
+            )),
```

Comment:
I find these inline/nested params a little hard to read/grok, but maybe that's just cos I'm less familiar with the patterns used here.

Reply:
No, it's good you mention this. I reduced the nesting a bit.